File: day10/puzzle.rb
# frozen_string_literal: true

input = File.readlines(*ARGV, chomp: true).map(&:split)

x = 1
x_at_each_cycle = []

input.each do |ins, v|
  x_at_each_cycle << x
  next if ins == "noop"

  x += v.to_i
  x_at_each_cycle << x
end

CYCLES_LIST = [220, 180, 140, 100, 60, 20].freeze

signal_strengths = CYCLES_LIST.sum { |cycle| cycle * x_at_each_cycle[cycle - 2] }

p signal_strengths # p1-> 17940

def draw_px_in_sprite?(mid_sprite, draw_px)
  mid_sprite - 1 <= draw_px and mid_sprite + 1 >= draw_px
end

CRT_ROWS = [200, 160, 120, 80, 40].freeze

flat_grid = Array.new(240) { "." }

crt = flat_grid.map.with_index do |px, cycle|
  height_correction = CRT_ROWS.find { |row| cycle > row - 1 } || 0
  draw_px = cycle - height_correction

  cycle_idx = [cycle - 1, 0].max
  x_val = x_at_each_cycle[cycle_idx]
  draw_px_in_sprite?(x_val, draw_px) ? "#" : px
end

crt.each_slice(40) { |line| puts line.join } # p2-> ZCBAJFJZ

File: day10/puzzle.js
const fs = require("fs");
const puzzleInput = fs.readFileSync(process.argv[2], { encoding: 'utf8' })
                      .split("\n")
                      .filter(Boolean)
                      .map((line) => line.split(" "));

let x = 1;
const xValueCollection = [];

for (const [instruction, value] of puzzleInput) {
  xValueCollection.push(x);
  if (instruction === "noop") continue;

  xValueCollection.push(x)
  x += Number(value);
}

CYCLE_LIST = [220, 180, 140, 100, 60, 20];
const signalStrengths = CYCLE_LIST.reduce((sum, cycle) => {
  return sum + cycle * xValueCollection[cycle];
}, 0)

console.log(signalStrengths); // Part 1

const drawPixelIsInSprite = (drawPx, sprite) => {
  return sprite - 1 <= drawPx && sprite + 1 >= drawPx;
}

const CRT_ROWS = [200, 160, 120, 80, 40];

const flatGrid = Array(240).fill(".");

const crt = flatGrid.map((px, cycle) => {
  const heightCorrection = CRT_ROWS.find((row) => cycle > row - 1) || 0;
  const drawPx = cycle - heightCorrection;

  const currentX = xValueCollection[cycle];
  return drawPixelIsInSprite(drawPx, currentX) ? "#" : px;
})

const crtImage = crt.reduce((output, px, index) => {
  return output + (index % 40 === 0 ? "\n" : "") + px;
})

console.log(crtImage); // Part 2
